### PR TITLE
build: use typescript-eslint@v6 internally

### PR DESCRIPTION
**Pull request by JoshuaKGoldberg**
### Does this PR add a new _feature_, or fix a _bug_?

Internal tooling. 😄 

## Checklist

- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [x] Is there an existing issue for this PR?
  - #4565
- [x] Have the files been linted and formatted?
- [x] Have the docs been updated to match the changes in the PR?
- [x] Have the tests been updated to match the changes in the PR?
- [x] Have you run the tests locally to confirm they pass?

## New Features

Upgraded typescript-eslint to v6, with reworked ESLint configurations.

You can read https://typescript-eslint.io/blog/announcing-typescript-eslint-v6-beta#user-facing-breaking-changes for the rationale behind the config changes. Essentially, the new recommended starter configs are:

* `"plugin:@typescript-eslint/recommended-type-checked"`
* `"plugin:@typescript-eslint/stylistic-type-checked"`

I've commented any changes ripe for discussion internally. This one was pretty straightforward compared to others though.

### CODE REVIEW

File: .eslintrc.cjs [hunk 1]
```diff
@@ -1,8 +1,17 @@
...
+  extends: [
+    'react-app',
+    'plugin:@typescript-eslint/recommended-type-checked',
+    'plugin:@typescript-eslint/stylistic-type-checked'
+  ],
 
   parser: '@typescript-eslint/parser',
 
+  parserOptions: {
+    project: true,
```
Comment (JoshuaKGoldberg) [inline .eslintrc.cjs:11]: [`project: true` is already in typescript-eslint@v5](https://typescript-eslint.io/linting/typed-linting). It's just a nice way to enable `parserOptions.project` 😄

File: .eslintrc.cjs [hunk 2]
```diff
@@ -31,6 +41,33 @@ module.exports = {
...
+    '@typescript-eslint/no-floating-promises': 'off',
+    '@typescript-eslint/no-unsafe-argument': 'off',
+    '@typescript-eslint/no-unsafe-assignment': 'off',
+    '@typescript-eslint/no-unsafe-call': 'off',
+    '@typescript-eslint/no-unsafe-member-access': 'off',
+    '@typescript-eslint/no-unsafe-return': 'off',
+    '@typescript-eslint/prefer-function-type': 'off',
+    '@typescript-eslint/sort-type-constituents': 'off',
+    '@typescript-eslint/unbound-method': 'off',
+    'prefer-rest-params': 'off',
```
Comment (JoshuaKGoldberg) [inline .eslintrc.cjs:68]: Each of these had at least a few existing warnings, so I didn't want to overreach and solve for them without asking. Are any of these rules ones you feel particularly strongly about?

* https://typescript-eslint.io/rules/ban-ts-comment
* https://typescript-eslint.io/rules/ban-types
* https://typescript-eslint.io/rules/consistent-indexed-object-style
* https://typescript-eslint.io/rules/consistent-type-definitions
* https://typescript-eslint.io/rules/no-confusing-void-expression
* https://typescript-eslint.io/rules/no-empty-function
* https://typescript-eslint.io/rules/no-floating-promises
* https://typescript-eslint.io/rules/no-unsafe-argument
* https://typescript-eslint.io/rules/no-unsafe-assignment
* https://typescript-eslint.io/rules/no-unsafe-call
* https://typescript-eslint.io/rules/no-unsafe-member-access
* https://typescript-eslint.io/rules/no-unsafe-return
* https://typescript-eslint.io/rules/prefer-function-type
* https://typescript-eslint.io/rules/sort-type-constituents
* https://typescript-eslint.io/rules/unbound-method
* https://eslint.org/docs/latest/rules/prefer-rest-params

If not, I can go ahead and fix for all their complaints.

File: yarn.lock
Comment (JoshuaKGoldberg) [inline yarn.lock:1]: 😅 what version of Yarn is this lockfile for? I couldn't figure it out locally...

Comment (markerikson) [reply yarn.lock:1]: We use Yarn 3.x across all our repos

Comment (JoshuaKGoldberg) [reply yarn.lock:1]: Ah, right, the `.yarn/releases` folder - thanks!

Comment (nickmccurdy) [reply yarn.lock:1]: Note that `yarn -v` or `yarn --version` should always show you the right version locally.

**Pull request closed without merging**